File: scripts/python/run_all_tests.py
#!/usr/bin/env python3
"""
Kaetram API Test Main Runner
"""
import os
import sys
import time
import argparse
from typing import List, Dict, Any

# Add utils path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'utils'))

from test_utils import TestConfig, Logger
from test_server_api import ServerAPITests
from test_hub_api import HubAPITests
from test_ai_agent import AIAgentFlowTests

def run_server_tests(config: TestConfig) -> Dict[str, Any]:
    """Run game server tests"""
    tests = ServerAPITests(config)
    tests.run_all_tests()
    return {
        "suite_name": "ServerAPI",
        "total_tests": len(tests.suite.results),
        "passed_tests": sum(1 for r in tests.suite.results if r.success),
        "failed_tests": sum(1 for r in tests.suite.results if not r.success),
        "duration": tests.suite.end_time - tests.suite.start_time if tests.suite.end_time else 0
    }

def run_hub_tests(config: TestConfig) -> Dict[str, Any]:
    """Run Hub tests"""
    tests = HubAPITests(config)
    tests.run_all_tests()
    return {
        "suite_name": "HubAPI",
        "total_tests": len(tests.suite.results),
        "passed_tests": sum(1 for r in tests.suite.results if r.success),
        "failed_tests": sum(1 for r in tests.suite.results if not r.success),
        "duration": tests.suite.end_time - tests.suite.start_time if tests.suite.end_time else 0
    }

def run_ai_agent_tests(config: TestConfig) -> Dict[str, Any]:
    """Run AI agent workflow tests"""
    tests = AIAgentFlowTests(config)
    tests.run_all_tests()
    return {
        "suite_name": "AIAgentFlow",
        "total_tests": len(tests.suite.results),
        "passed_tests": sum(1 for r in tests.suite.results if r.success),
        "failed_tests": sum(1 for r in tests.suite.results if not r.success),
        "duration": tests.suite.end_time - tests.suite.start_time if tests.suite.end_time else 0
    }

def print_summary(results: List[Dict[str, Any]], total_duration: float):
    """Print test summary"""
    print("\n" + "="*80)
    print("🎮 Kaetram API Test Summary")
    print("="*80)
    
    total_tests = sum(r['total_tests'] for r in results)
    total_passed = sum(r['passed_tests'] for r in results)
    total_failed = sum(r['failed_tests'] for r in results)
    
    for result in results:
        suite_name = result['suite_name'].ljust(20)
        status = "✅ PASSED" if result['failed_tests'] == 0 else "❌ FAILED"
        test_count = f"{result['passed_tests']}/{result['total_tests']}"
        duration = f"{result['duration']:.2f}s"
        
        print(f"{suite_name} | {status.ljust(9)} | {test_count.ljust(8)} | {duration.ljust(8)}")
    
    print("-"*80)
    total_status = "✅ PASSED" if total_failed == 0 else "❌ FAILED"
    total_count = f"{total_passed}/{total_tests}"
    total_dur = f"{total_duration:.2f}s"
    
    print(f"{'TOTAL'.ljust(20)} | {total_status.ljust(9)} | {total_count.ljust(8)} | {total_dur.ljust(8)}")
    
    if total_failed > 0:
        print(f"\n⚠️  {total_failed} tests failed. Check the logs for details.")
    else:
        print(f"\n🎉 All tests passed! ({total_tests} tests)")
    
    print("="*80)

def main():
    """Main function"""
    parser = argparse.ArgumentParser(description="Kaetram API Test Runner")
    parser.add_argument("--config", default="config.json", help="Configuration file path")
    parser.add_argument("--server", action="store_true", help="Run only server API tests")
    parser.add_argument("--hub", action="store_true", help="Run only Hub API tests")
    parser.add_argument("--ai", action="store_true", help="Run only AI agent workflow tests")
    parser.add_argument("--verbose", action="store_true", help="Enable verbose output")
    parser.add_argument("--parallel", action="store_true", help="Run tests in parallel")
    parser.add_argument("--cleanup", action="store_true", help="Clean up test data after execution")
    
    args = parser.parse_args()
    
    try:
        # Load configuration
        config = TestConfig(args.config)
        logger = Logger(config)
        
        # Override config with command line args
        if args.verbose:
            logger.logger.setLevel("DEBUG")
        
        # Determine which tests to run
        run_server = args.server or not (args.hub or args.ai)
        run_hub = args.hub or not (args.server or args.ai)
        run_ai = args.ai or not (args.server or args.hub)
        
        # If specific tests are requested, only run those
        if args.server or args.hub or args.ai:
            run_server = args.server
            run_hub = args.hub
            run_ai = args.ai
        
        results = []
        start_time = time.time()
        
        print("🚀 Starting Kaetram API Tests")
        print(f"Configuration: {args.config}")
        print(f"Server URL: {config.get('server.base_url')}")
        print(f"Hub URL: {config.get('hub.base_url')}")
        print("-"*50)
        
        # Run tests
        if run_server:
            print("Running Server API Tests...")
            try:
                result = run_server_tests(config)
                results.append(result)
                logger.info(f"Server API tests completed: {result['passed_tests']}/{result['total_tests']} passed")
            except Exception as e:
                logger.error(f"Server API tests failed: {e}")
                results.append({
                    "suite_name": "ServerAPI",
                    "total_tests": 0,
                    "passed_tests": 0,
                    "failed_tests": 1,
                    "duration": 0
                })
        
        if run_hub:
            print("Running Hub API Tests...")
            try:
                result = run_hub_tests(config)
                results.append(result)
                logger.info(f"Hub API tests completed: {result['passed_tests']}/{result['total_tests']} passed")
            except Exception as e:
                logger.error(f"Hub API tests failed: {e}")
                results.append({
                    "suite_name": "HubAPI",
                    "total_tests": 0,
                    "passed_tests": 0,
                    "failed_tests": 1,
                    "duration": 0
                })
        
        if run_ai:
            print("Running AI Agent Workflow Tests...")
            try:
                result = run_ai_agent_tests(config)
                results.append(result)
                logger.info(f"AI Agent tests completed: {result['passed_tests']}/{result['total_tests']} passed")
            except Exception as e:
                logger.error(f"AI Agent tests failed: {e}")
                results.append({
                    "suite_name": "AIAgentFlow",
                    "total_tests": 0,
                    "passed_tests": 0,
                    "failed_tests": 1,
                    "duration": 0
                })
        
        # Calculate total duration
        total_duration = time.time() - start_time
        
        # Print summary
        print_summary(results, total_duration)
        
        # Save combined report
        combined_report = {
            "timestamp": time.time(),
            "total_duration": total_duration,
            "suites": results,
            "summary": {
                "total_tests": sum(r['total_tests'] for r in results),
                "passed_tests": sum(r['passed_tests'] for r in results),
                "failed_tests": sum(r['failed_tests'] for r in results),
                "success_rate": (sum(r['passed_tests'] for r in results) / sum(r['total_tests'] for r in results) * 100) if sum(r['total_tests'] for r in results) > 0 else 0
            }
        }
        
        # Save report
        import json
        from pathlib import Path
        
        report_dir = Path("test_reports")
        report_dir.mkdir(exist_ok=True)
        
        timestamp = time.strftime("%Y%m%d_%H%M%S")
        report_file = report_dir / f"combined_report_{timestamp}.json"
        
        with open(report_file, 'w', encoding='utf-8') as f:
            json.dump(combined_report, f, indent=2, ensure_ascii=False, default=str)
        
        print(f"\n📊 Combined report saved to: {report_file}")
        
        # Exit with error code if any tests failed
        total_failed = sum(r['failed_tests'] for r in results)
        if total_failed > 0:
            sys.exit(1)
        
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main() 

File: scripts/javascript/run_all_tests.js
#!/usr/bin/env node
/**
 * Kaetram API Test Main Runner - JavaScript Version
 */
const { program } = require('commander');
const chalk = require('chalk');
const { TestConfig } = require('../utils/test_utils');
const ServerAPITests = require('./test_server_api');

async function runServerTests(config) {
    console.log(chalk.blue('Running game server API tests...'));
    const tests = new ServerAPITests(config);
    await tests.runAllTests();
    
    return {
        suite_name: 'ServerAPI',
        total_tests: tests.suite.results.length,
        passed_tests: tests.suite.results.filter(r => r.success).length,
        failed_tests: tests.suite.results.filter(r => !r.success).length,
        duration: (tests.suite.endTime - tests.suite.startTime) / 1000
    };
}

function printSummary(results, totalDuration) {
    console.log('\n' + '='.repeat(80));
    console.log(chalk.bold('🎮 Kaetram API Test Summary'));
    console.log('='.repeat(80));
    
    const totalTests = results.reduce((sum, r) => sum + r.total_tests, 0);
    const totalPassed = results.reduce((sum, r) => sum + r.passed_tests, 0);
    const totalFailed = results.reduce((sum, r) => sum + r.failed_tests, 0);
    
    results.forEach(result => {
        const suiteName = result.suite_name.padEnd(20);
        const status = result.failed_tests === 0 ? 
            chalk.green('✅ PASSED') : chalk.red('❌ FAILED');
        const testCount = `${result.passed_tests.toString().padStart(3)}/${result.total_tests.toString().padStart(3)}`;
        const duration = `${result.duration.toFixed(2)}s`.padStart(8);
        
        console.log(`${suiteName} | ${status} | ${testCount} | ${duration}`);
    });
    
    console.log('-'.repeat(80));
    const totalStatus = totalFailed === 0 ? 
        chalk.green('✅ PASSED') : chalk.red('❌ FAILED');
    const totalCount = `${totalPassed.toString().padStart(3)}/${totalTests.toString().padStart(3)}`;
    const totalDur = `${totalDuration.toFixed(2)}s`.padStart(8);
    
    console.log(`${'TOTAL'.padEnd(20)} | ${totalStatus} | ${totalCount} | ${totalDur}`);
    
    if (totalFailed > 0) {
        console.log(chalk.red(`\n⚠️  ${totalFailed} tests failed. Check the logs for details.`));
    } else {
        console.log(chalk.green(`\n🎉 All tests passed! (${totalTests} tests)`));
    }
    
    console.log('='.repeat(80));
}

async function main() {
    program
        .name('kaetram-api-test')
        .description('Kaetram API Test Runner')
        .version('1.0.0')
        .option('--config <file>', 'Configuration file path', 'config.json')
        .option('--server', 'Run only server API tests')
        .option('--hub', 'Run only Hub API tests')
        .option('--ai', 'Run only AI agent workflow tests')
        .option('--verbose', 'Enable verbose output')
        .option('--parallel', 'Run tests in parallel')
        .option('--cleanup', 'Clean up test data after execution')
        .parse();

    const options = program.opts();

    try {
        // Load configuration
        const config = new TestConfig(options.config);
        
        // Determine which tests to run
        let runServer = options.server || (!options.hub && !options.ai);
        let runHub = options.hub || (!options.server && !options.ai);
        let runAi = options.ai || (!options.server && !options.hub);
        
        // If specific tests are requested, only run those
        if (options.server || options.hub || options.ai) {
            runServer = options.server;
            runHub = options.hub;
            runAi = options.ai;
        }

        const results = [];
        const startTime = Date.now();

        console.log(chalk.bold('🚀 Starting Kaetram API Tests'));
        console.log(`Configuration: ${options.config}`);
        console.log(`Server URL: ${config.get('server.base_url')}`);
        console.log(`Hub URL: ${config.get('hub.base_url')}`);
        console.log('-'.repeat(50));

        // Run tests
        if (runServer) {
            console.log(chalk.blue('Running Server API Tests...'));
            try {
                const result = await runServerTests(config);
                results.push(result);
                console.log(chalk.green(`Server API tests completed: ${result.passed_tests}/${result.total_tests} passed`));
            } catch (error) {
                console.error(chalk.red(`Server API tests failed: ${error.message}`));
                results.push({
                    suite_name: 'ServerAPI',
                    total_tests: 0,
                    passed_tests: 0,
                    failed_tests: 1,
                    duration: 0
                });
            }
        }

        if (runHub) {
            console.log(chalk.blue('Running Hub API Tests...'));
            try {
                // Note: Hub tests would be implemented here
                console.log(chalk.yellow('Hub API tests not yet implemented in JavaScript version'));
                results.push({
                    suite_name: 'HubAPI',
                    total_tests: 0,
                    passed_tests: 0,
                    failed_tests: 0,
                    duration: 0
                });
            } catch (error) {
                console.error(chalk.red(`Hub API tests failed: ${error.message}`));
                results.push({
                    suite_name: 'HubAPI',
                    total_tests: 0,
                    passed_tests: 0,
                    failed_tests: 1,
                    duration: 0
                });
            }
        }

        if (runAi) {
            console.log(chalk.blue('Running AI Agent Workflow Tests...'));
            try {
                // Note: AI agent tests would be implemented here
                console.log(chalk.yellow('AI Agent workflow tests not yet implemented in JavaScript version'));
                results.push({
                    suite_name: 'AIAgentFlow',
                    total_tests: 0,
                    passed_tests: 0,
                    failed_tests: 0,
                    duration: 0
                });
            } catch (error) {
                console.error(chalk.red(`AI Agent tests failed: ${error.message}`));
                results.push({
                    suite_name: 'AIAgentFlow',
                    total_tests: 0,
                    passed_tests: 0,
                    failed_tests: 1,
                    duration: 0
                });
            }
        }

        // Calculate total duration
        const totalDuration = (Date.now() - startTime) / 1000;

        // Print summary
        printSummary(results, totalDuration);

        // Save combined report
        const fs = require('fs');
        const path = require('path');
        
        const reportDir = path.join(__dirname, '..', 'test_reports');
        if (!fs.existsSync(reportDir)) {
            fs.mkdirSync(reportDir, { recursive: true });
        }

        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const reportFile = path.join(reportDir, `combined_report_${timestamp}.json`);

        fs.writeFileSync(reportFile, JSON.stringify(combinedReport, null, 2));
        console.log(chalk.cyan(`\n📊 Combined report saved to: ${reportFile}`));

        // Exit with error code if any tests failed
        const totalFailed = results.reduce((sum, r) => sum + r.failed_tests, 0);
        if (totalFailed > 0) {
            process.exit(1);
        }

    } catch (error) {
        console.error(chalk.red(`Error: ${error.message}`));
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = { runServerTests, printSummary }; 